### ISSUE TITLE
False positive from prefer_const_constructors

### ISSUE THREAD

**Issue by kentcb** [dart-lang/linter#1022]
This code:
```dart
class TypeWrapper {
  final Type type;

  const TypeWrapper(this.type);
}

void method<T>() {
  final typeWrapper = new TypeWrapper(T);
}
```
Triggers `prefer_const_constructors`:
```
Prefer const with constant constructors.
```
However, attempting to "fix" per the suggestion with this:
```dart
const typeWrapper = const TypeWrapper(T);
```
Yields a compiler error:
```
compiler message: lib/utility/repro.dart:82:41: Error: Not a constant expression.
compiler message:   const typeWrapper = const TypeWrapper(T);
compiler message:                                         ^
compiler message: lib/utility/repro.dart:82:41: Error: Type variable 'T' can't be used as a constant expression #lib1::method::T.
compiler message:   const typeWrapper = const TypeWrapper(T);
```

**Comment by zoechi** [comment]
I copied your code, but can't reproduce. 
Have you tried restarting your IDE?
What Dart version are you using?

**Comment by kentcb** [comment]
Note that I stuffed the code up in my original post because I had it as `const` instead of `final`. I've updated that.

I also tried restarting VS Code and I still get the analysis warning. I also still get the build error when attempting to run.

`flutter doctor -v`:
```
[√] Flutter (Channel master, v0.5.3-pre.9, on Microsoft Windows [Version 10.0.17134.48], locale en-US)
    • Flutter version 0.5.3-pre.9 at C:\Users\kent\Repository\flutter
    • Framework revision 0e4770e809 (4 days ago), 2018-06-07 18:27:08 -0700
    • Engine revision fca976d8c7
    • Dart version 2.0.0-dev.60.0.flutter-a5e41681e5

[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at C:\Android\android-sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • ANDROID_HOME = C:\Android\android-sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b02)
    • All Android licenses accepted.

[√] Android Studio (version 3.1)
    • Android Studio at C:\Program Files\Android\Android Studio
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b02)

[√] IntelliJ IDEA Community Edition (version 2017.3)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2017.3.2
    • Flutter plugin version 22.2.2
    • Dart plugin version 173.4548.30

[!] VS Code, 64-bit edition (version 1.23.1)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[√] Connected devices (1 available)
    • Android SDK built for x86 64 • emulator-5554 • android-x64 • Android 8.0.0 (API 26) (emulator)

! Doctor found issues in 1 category.
```

**Comment by zoechi** [comment]
I can reproduce now, but I think it's working as intended. 
I think this is exactly what the lint was made for.
Why don't you want to use `const` here?

**Comment by zoechi** [comment]
I can reproduce now and I see the problem. The methods generic type parameter `T` is not const.

**Comment by a14n** [comment]
related to dart-lang/sdk#33612

**Comment by pq** [comment]
Also see: #628

**Comment by pq** [comment]
Migrated from https://github.com/dart-lang/sdk/issues/33289, where @gspencergoog reports:


> If I have:
> ```dart
> class A<T> {
>   const A();
> }
> ```
> 
> Then this code:
> ```dart
> void foo<T>() {
>   final Type type = new A<T>().runtimeType;
> }
> ```
> 
> If I have `prefer_const_constructors` turned on, the analyzer thinks that the expression should be const, since the constructor is const, but if I make it const and try to compile this code, the compiler says it's not a const expression (since it is a generic with a variable type).
> 
> ```
> compiler message: lib/toggle_controller.dart:9:29: Error: Not a constant expression.
> compiler message:   final Type type = const A<T>().runtimeType;
> compiler message:                             ^
> compiler message: lib/toggle_controller.dart:9:29: Error: Type variable 'T' can't be used as a constant expression #lib1::foo::T.
> compiler message:   final Type type = const A<T>().runtimeType;
> compiler message: 
> ```
> 
> ```
> Dart VM version: 2.0.0-dev.59.0.flutter-ff815d05a5 (Tue May 29 20:01:09 2018 +0000) on "macos_x64"
> ```
>

**Comment by scheglov** [comment]
I think https://github.com/dart-lang/linter/pull/1159 fixed the issue.
